url27a7ae3508_serkan_ozal_ocean_of_memories_tgz-pJ8-com_zeroturnaround_rebellabs_oceanofmemories_article1_objectlayout_ObjectMemoryLayoutDemoJ8/src/com/zeroturnaround/rebellabs/oceanofmemories/common/util/Util.java

```diff
@@ -0,0 +1,88 @@
+package com.zeroturnaround.rebellabs.oceanofmemories.common.util;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URLDecoder;
+
+public class Util {
+	
+	public static final int ASCII_CODE_OF_MIN_PRINTABLE_CHARACTER = 33;
+	public static final int ASCII_CODE_OF_MAX_PRINTABLE_CHARACTER = 127;
+	
+	private Util() {
+		
+	}
+	
+	public static double makeDouble2DigitPrecisioned(double value) {
+		return ((int)(value * 100) / (double)100);
+	}
+	
+	public static double percantegeOf(double value1, double value2) {
+		return makeDouble2DigitPrecisioned((value1 * 100) / value2);
+	}
+
+	public static boolean isPrintableChacter(char c) {
+		return 
+			c >= ASCII_CODE_OF_MIN_PRINTABLE_CHARACTER && 
+			c <= ASCII_CODE_OF_MAX_PRINTABLE_CHARACTER;
+	}
+	
+	public static void waitFor(long millisecods) {
+		try {
+			Thread.sleep(millisecods);
+		} 
+		catch (InterruptedException e) {
+			e.printStackTrace();
+		}
+	}
+	
+	@SuppressWarnings("deprecation")
+	public static void runClass(Class<?> classHasMainMethod, String[] vmArguments, String[] programArguments) {
+		try {
+			StringBuilder sbVmArguments = new StringBuilder();
+			StringBuilder sbProgramArguments = new StringBuilder();
+			
+			if (vmArguments != null) {
+				for (String arg : vmArguments) {
+					sbVmArguments.append(arg).append(" ");
+				}
+			}
+			
+			if (programArguments != null) {
+				for (String arg : programArguments) {
+					sbProgramArguments.append(arg).append(" ");
+				}
+			}
+			
+			String classPath = URLDecoder.decode(Util.class.getClassLoader().getResource(".").getPath());
+			Process proc = 
+					Runtime.getRuntime().exec(
+							"java" + " -cp " + "\"" + classPath + "\"" + " " +
+							sbVmArguments.toString() + " " +
+							classHasMainMethod.getName() + " " +
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;classHasMainMethod&#39; (index: 0 | type: Class). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;vmArguments&#39;, &#39;programArguments&#39;. The same parameter type is validated here for example: url27a7ae3508_serkan_ozal_ocean_of_memories_tgz-pJ8-com_zeroturnaround_rebellabs_oceanofmemories_article1_objectlayout_ObjectMemoryLayoutDemoJ8/src/com/zeroturnaround/rebellabs/oceanofmemories/common/util/ClassUtil.java:33. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).